### PR TITLE
Add Blog post gcp-workload-identity-with-gke-and-terraform

### PR DESCRIPTION
**Pull request by parabolic**

### CODE REVIEW

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+  <a href="https://trixelized.itch.io/">
+    <img alt="pixel_art_forest" title="Pixel Art Forest" src="assets/images/2022_03_pixelized_forest.jpg" style="border:1px solid #2C4D56" height="460">
+  </a>
+</p>
+<br/>
+
+> "The only truly secure system is one that is powered off, cast in a block of concrete and sealed in a lead-lined room with armed guards." - Gene Spafford
+
+<br/>
+In this blog post, I will write about increasing the security posture. It centers on running Google Kubernetes Engine (GKE) workloads on the Google Cloud Platform (GCP). The focus falls mainly on authentication. That is the section between the Identity and Access Management (IAM) service and the GKE workloads. Foremost, I need to set the stage properly.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:21]:
```suggestion
In this blog post, I will write about improving one's security posture by running Google Kubernetes Engine (GKE) workloads on the Google Cloud Platform (GCP). My focus falls mainly on authentication: the section between the Identity and Access Management (IAM) service and the GKE workloads. But first, I need to set the stage.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+## [Service Accounts, IAM Roles, And Key Pairs]
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:23]:
```suggestion
## [Service Accounts, IAM Roles, and Key Pairs]
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+#### [Service Accounts]
+Service accounts are types of accounts typically used by applications or so-called workloads.
+With a service account, the application can authenticate to other GCP resources or APIs. It is an entity defined by GCP and resides in the IAM service. Its email address representation e.g. `sa-name@gcp-project-id.iam.gserviceaccount.com` is unique to a given GCP account.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:27]:
```suggestion
With a service account, the application can authenticate to other GCP resources or APIs. It is an entity defined by GCP and resides in the IAM service. Its email address representation, e.g. `sa-name@gcp-project-id.iam.gserviceaccount.com`, is unique to a given GCP account.
```
```suggestion
With a service account, the application can authenticate to other GCP resources or APIs. It is an entity defined by GCP and resides in the IAM service. Its email address representation, e.g. `sa-name@gcp-project-id.iam.gserviceaccount.com`, is unique to a given GCP account.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+#### [IAM Roles]
+A service account usually is granted an IAM role that defines the authorization scope for other GCP resources.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:30]:
```suggestion
A service account is usually granted an IAM role that defines the authorization scope for other GCP resources.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+With that said, we can safely assume that a service account deals with the authentication and the IAM role with the authorization.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:31]:
```suggestion
With that said, we can safely assume that a service account deals with authentication and the IAM role deals with authorization.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+Each service account correlates with a public/private RSA key pair.
+Depending on how the private key is generated, it diverges into two types:
+- Google-managed key pairs.
+- User-managed key pairs.
+
+###### [Google-Managed Key pairs]
+Are managed by Google, and only the public key can be viewed and verified.
+The most common use of such keys is for generating short-lived credentials.
+They are automatically rotated and used for signing and are valid for a maximum of two weeks.
+Services like App Engine and Compute Engine are using Google-managed key pairs.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:43]:
```suggestion
Services like App Engine and Compute Engine use Google-managed key pairs.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+A service account usually is granted an IAM role that defines the authorization scope for other GCP resources.
+With that said, we can safely assume that a service account deals with the authentication and the IAM role with the authorization.
+
+#### [Public-Private Key Pairs]
+Each service account correlates with a public/private RSA key pair.
+Depending on how the private key is generated, it diverges into two types:
+- Google-managed key pairs.
+- User-managed key pairs.
+
+###### [Google-Managed Key pairs]
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:39]:
```suggestion
###### [Google-Managed Key Pairs]
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+Are managed by Google, and only the public key can be viewed and verified.
+The most common use of such keys is for generating short-lived credentials.
+They are automatically rotated and used for signing and are valid for a maximum of two weeks.
+Services like App Engine and Compute Engine are using Google-managed key pairs.
+They are my preferred method (it should be everyone's default). That is because I do not have to care about managing or storing any private keys which are considered sensitive data.
+
+###### [User-Managed Key pairs]
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:46]:
```suggestion
###### [User-Managed Key Pairs]
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+  "client_email": "cloudlad@some-gcp-project-id.iam.gserviceaccount.com",
+  "client_id": "30348941921832193121",
+  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+  "token_uri": "https://oauth2.googleapis.com/token",
+  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/cloudlad%40some-gcp-project-id.iam.gserviceaccount.com"
+}
+```
+<p align="center">
+  <b style="font-size:0.7vw" >Figure 1. A user-managed bogus service account key.</b>
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:69]:
```suggestion
  <b style="font-size:0.7vw" >Figure 1. A bogus user-managed service account key.</b>
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+</p>
+
+This User-managed key by default has no expiration date. It also inherits the same permissions as the service account.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:72]:
```suggestion
This user-managed key by default has no expiration date. It also inherits the same permissions as the service account.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+Naturally, such keys are considered a security risk. The responsibility for them falls on the user. See Figure 2.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:73]:
```suggestion
Naturally, such keys are considered a security risk. The responsibility to keep them safe falls on the user. See Figure 2.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+<p align="center">
+    <img alt="service account key" title="service account key" src="assets/images/2022_03_service_account_key.png" style="border:1px solid #2C4D56">
+</p>
+
+<p align="center">
+  <b style="font-size:0.7vw" >Figure 2. Places where a service account key might exist.</b>
+</p>
+
+
+## [Why use service account keys?]
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:85]:
```suggestion
## [Why Use Service Account Keys?]
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+
+One of the more common use cases for service accounts is for accessing resources in GCP, be it a Compute Engine VM or a GKE node pool.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:87]:
```suggestion
One of the more common use cases for service accounts is accessing resources in GCP, be it a Compute Engine VM or a GKE node pool.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+For clarification, a node pool is a group of VMs sharing the same configuration.
+
+
+The pitfall here is, in the case of the GKE node pools, all of the pods on the nodes will inherit the permissions assigned to the service account.
+Let's assume that application A needs read-only access to PubSub, and application B needs write access to Memorystore (Redis).
+The applications are both running as pods on nodes in a node pool.
+If we attach a service account on the node pool level, both applications will inherit the permissions from the attached service account.  Both will have unnecessarily broad permissions, increasing the attack surface and risk.
+The example is only with two applications.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:95]:
```suggestion
And take note: this example is only with two applications.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+A GKE cluster hosts way more, not counting the system pods.
+
+The solution is rather obvious, do not assign permissions on the node pool level.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:98]:
```suggestion
The solution is rather obvious: do **not** assign permissions on the node pool level.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+
+If I cannot do it on the node level, I would have to do it on the application level. E.g., secrets loaded directly in the pod.
+The secrets API object in K8s holds a small amount of sensitive data. The data gets put into a running pod or container image, either mounted or exposed as an environment variable.
+
+With this, I am back to square one, user-generated service account keys loaded into pods.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:103]:
```suggestion
With this, I am back to square one: user-generated service account keys loaded into pods.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+
+## [Workload Identity]
+
+Workload Identity is a process that enables workloads to impersonate (IAM) service accounts to access GCP services.
+The workloads might or might not be running on GCP.
+In simpler terms, the workloads assume the authorization scope of a service account and use short-lived credentials that the user does not need to manage in any way. Effectively preventing the leakage of credentials and decreasing the possible attack surface. See Figure 3.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:109]:
```suggestion
In simpler terms, the workloads assume the authorization scope of a service account and use short-lived credentials that the user does not need to manage in any way. This effectively preventis the leakage of credentials and decreases the possible attack surface. See Figure 3.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+
+<p align="center">
+  <img alt="workload-identity-level-3" title="Workload Identity Level 3" src="assets/images/2022_03_level_3.png" style="border:1px solid #2C4D56">
+</p>
+
+<p align="center">
+  <b style="font-size:0.7vw" >Figure 4. Workload Identity.</b>
+</p>
+
+## [Practical Example In Terraform]
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:144]:
```suggestion
## [Practical Example in Terraform]
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+</p>
+
+<p align="center">
+  <b style="font-size:0.7vw" >Figure 5. Simplified representation of the cloud resources created by the Terraform root module.</b>
+</p>
+
+For more details about the module and its configuration, refer to the [README.md].
+
+I would have preferred to create all of the resources in one go, but unfortunately, it is not possible with Terraform’s CRUD cycle. The problem is that the K8s provider cannot authenticate against the GKE cluster in one apply run. That makes perfect sense because the GKE cluster is not present in the application stage, and the K8s provider cannot authenticate.
+To programmatically create all of the resources, I utilize the [local_file] provider to generate a helper bash script from a Terraform template. The bash script `k8s.sh` will appear in the working directory of the root module after a successful apply run. It can be [overiden to any valid path] in the Operating System.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:173]:
```suggestion
To programmatically create all of the resources, I use the [local_file] provider to generate a helper bash script from a Terraform template. The bash script `k8s.sh` will appear in the working directory of the root module after a successful apply run. It can be [overirden to any valid path] in the Operating System.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+    },
+    "secrets": [
+        {
+            "name": "cloudlad-token-thlzr"
+        }
+    ]
+}
+```
+<p align="center">
+  <b style="font-size:0.7vw" >Figure 7. Service Account annotation got with kubectl.</b>
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:238]:
```suggestion
  <b style="font-size:0.7vw" >Figure 7. Service Account annotation with kubectl.</b>
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+</p>
+
+You might ask yourself, "OK, What now?". Checking whether the coupling between GCP IAM and GKE works is straightforward. The newly created pod runs indefinite sleep. And the container is started from the official GCP SDK image, meaning the gcloud CLI is readily available.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:241]:
```suggestion
You might ask yourself, "OK, what now?" Checking whether the coupling between GCP IAM and GKE works is straightforward. The newly created pod runs indefinite sleep. And the container is started from the official GCP SDK image, meaning the gcloud CLI is readily available.
```

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+
+I will assume a shell in the running container. Then check the available service accounts, and finally, create a PubSub topic. It is all possible because the pod inherits the [roles/pubsub.admin] role. See Figure 8.1.
```
Comment (vasstambo) [inline _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:243]:
```suggestion
I will assume a shell is in the running container. Then check the available service accounts, and finally, create a PubSub topic. It is all possible because the pod inherits the [roles/pubsub.admin] role. See Figure 8.1.
```

**Review by vasstambo — CHANGES_REQUESTED** [review]

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown (hunk 19 above)
Comment (parabolic) [reply _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:177]: I will just change utilize to use, the other word looks like a unintentional typo.

File: _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown (hunk 22 above)
Comment (parabolic) [reply _posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown:247]: I will change the wording a bit here. By assume I meant execute a shell.

**Comment by parabolic** [comment]
Thank you so much @vasstambo. I will be merging this one :ship:
